batch2/batch/batch.py

```diff
@@ -1,19 +1,1278 @@
+async def index(request, userdata):
+    location = request.app.router['batches'].url_for()
+    raise web.HTTPFound(location=location)
+
+
+async def update_job_with_pod(job, pod):  # pylint: disable=R0911
+    log.info(f'update job {job.id if job else "None"} with pod {pod.metadata.name if pod else "None"}')
+    if job and job._state == 'Pending':
+        if pod:
+            log.error(f'job {job.id} has pod {pod.metadata.name}, ignoring')
+        return
+
+    if pod and (not job or job.is_complete()):
+        err = await app['driver'].delete_pod(name=pod.metadata.name)
+        if err is not None:
+            traceback.print_tb(err.__traceback__)
+            log.info(f'failed to delete pod {pod.metadata.name} for job {job.id if job else "None"} due to {err}, ignoring')
+        return
+
+    if job and job._cancelled and not job.always_run and job._state == 'Running':
+        await job.set_state('Cancelled')
+        await job._delete_pod()
+        return
+
+    if pod and pod.status and pod.status.phase == 'Pending':
+        all_container_statuses = pod.status.container_statuses or []
+        for container_status in all_container_statuses:
+            if (container_status.state and
+                    container_status.state.waiting and
+                    container_status.state.waiting.reason):
+                await job.mark_complete(pod)
+                return
+
+    if not pod:
+        log.info(f'job {job.id} no pod found, rescheduling')
+        await job.mark_unscheduled()
+        return
+
+    if pod and pod.status and pod.status.reason == 'Evicted':
+        POD_EVICTIONS.inc()
+        log.info(f'job {job.id} mark unscheduled -- pod was evicted')
+        await job.mark_unscheduled()
+        return
+
+    if pod and pod.status and pod.status.phase in ('Succeeded', 'Failed'):
+        log.info(f'job {job.id} mark complete')
+        await job.mark_complete(pod)
+        return
+
+    if pod and pod.status and pod.status.phase == 'Unknown':
+        log.info(f'job {job.id} mark unscheduled -- pod phase is unknown')
+        await job.mark_unscheduled()
+        return
+
+
+class DeblockedIterator:
+    def __init__(self, it):
+        self.it = it
+
+    def __aiter__(self):
+        return self
+
+    def __anext__(self):
+        return blocking_to_async(app['blocking_pool'], self.it.__next__)
+
+
+async def pod_changed(pod):
+    job = await Job.from_pod(pod)
+    await update_job_with_pod(job, pod)
+
+
+async def refresh_pods():
+    log.info(f'refreshing pods')
+
+    pod_jobs = [Job.from_record(record) for record in await db.jobs.get_records_where({'state': 'Running'})]
+
+    pods = app['driver'].list_pods()
+    log.info(f'batch had {len(pods)} pods')
+
+    seen_pods = set()
+    for pod_dict in pods:
+        pod = v1.api_client._ApiClient__deserialize(pod_dict, kube.client.V1Pod)
+        pod_name = pod.metadata.name
+        seen_pods.add(pod_name)
+        asyncio.ensure_future(pod_changed(pod))
```

Comment:
If we want to run all the pod_changed in parallel we should use `asyncio.gather` instead of `ensure_future`. Gather allows us to check if any coroutine failed and strictly bounds what and where the concurrency occurs which seems easier to reason about than full concurrency.